File: CompiladorLC/CompiladorLC.cpp
#include "Estado.hpp"
#include "GeradorDeAutomatos.hpp"
#include <fstream>

void imprimirVectorDeStrings(std::vector<std::string> vetor) {
    for (std::string linha : vetor) {
        std::cout << linha << std::endl;
    }
}

std::vector<std::string> lerArquivo(std::string nomeArquivo) {
    std::ifstream arquivo(nomeArquivo);
    std::vector<std::string> linhas;
    std::string linha;
    while (std::getline(arquivo, linha)) {
        linhas.push_back(linha);
    }
    return linhas;
}

void escreveArquivo(std::string nomeArquivo, std::vector<std::string> linhas) {
    std::ofstream arquivo(nomeArquivo);
    for (std::string linha : linhas) {
        arquivo << linha << std::endl;
    }
}

std::vector<std::string> preProcessamento(std::vector<std::string> linhas) {
    std::vector<std::string> linhasPreProcessadas;
    for (std::string linha : linhas) {
        std::string linhaPreProcessada;
        for (char c : linha) {
            if (c == ' ') {
                continue;
            }
            linhaPreProcessada += c;
        }
        linhasPreProcessadas.push_back(linhaPreProcessada);
    }
    return linhasPreProcessadas;
}

bool verificarVersaodoSO() {
    #ifdef _WIN32
	    return true;
    #else
	    return false;
    #endif
}

void corrigirCodificacaoDoTerminal(bool versaoWindows) {
	if (versaoWindows) {
		system("chcp 65001");
	}
}

void ajustesIniciais() {
	corrigirCodificacaoDoTerminal(verificarVersaodoSO());
}

int main()
{
	ajustesIniciais();

    std::cout << "Hello World!\n";
    std::vector<std::string> linhas = lerArquivo("C:\\Users\\mathe\\Downloads\\Novo(a) Documento de Texto.txt");
    std::cout << "Conteúdo do arquivo lido: ";
	imprimirVectorDeStrings(linhas);

	std::vector<std::string> regexs = { "a", "b", "c" };
    Automato LCAutomato("My Automato", regexs); //q0

    std::vector<std::string> regexs2 = { "d", "e", "f" }; //q1
    std::vector<std::string> regexs3 = { "g", "h", "i" }; //q2

    LCAutomato.addEstado(regexs2, TipoEstado::NORMAL);
    LCAutomato.addEstado(regexs3, TipoEstado::FINAL);

    LCAutomato.addTransicaoEntreEstados("q0", "q1", std::vector<std::string>{"z"});
    LCAutomato.addTransicaoEntreEstados("q1", "q2", std::vector<std::string>{"z"});

    std::cout << LCAutomato.transitarEntreEstados("z") << std::endl;
    std::cout << LCAutomato.transitarEntreEstados("z") << std::endl;

    //LCAutomato.~Automato();
    
    return 0;
}

File: CompiladorLC-traduzido/CompiladorLC.py
import os
from Automato_.GeradorDeAutomatos import Automato
from Automato_.Estado import TipoEstado


def imprimir_lista_de_strings(lista):
    """Imprime uma lista de strings."""
    for linha in lista:
        print(linha)


def ler_arquivo(nome_arquivo):
    """Lê um arquivo e retorna uma lista de linhas."""
    try:
        with open(nome_arquivo, "r", encoding="utf-8") as arquivo:
            return arquivo.readlines()
    except FileNotFoundError:
        print(f"Erro: Arquivo '{nome_arquivo}' não encontrado.")
        return []


def escreve_arquivo(nome_arquivo, linhas):
    """Escreve uma lista de strings em um arquivo."""
    with open(nome_arquivo, "w", encoding="utf-8") as arquivo:
        for linha in linhas:
            arquivo.write(linha + "\n")


def pre_processamento(linhas):
    """Remove espaços em branco de cada linha."""
    return ["".join(linha.split()) for linha in linhas]


def verificar_versao_do_so():
    """Verifica se o sistema operacional é Windows."""
    return os.name == "nt"


def corrigir_codificacao_do_terminal(versao_windows):
    """Ajusta a codificação do terminal no Windows."""
    if versao_windows:
        os.system("chcp 65001")


def ajustes_iniciais():
    """Executa ajustes iniciais no terminal."""
    corrigir_codificacao_do_terminal(verificar_versao_do_so())


def main():
    """Função principal."""
    ajustes_iniciais()

    print("Hello World!")
    
    # Ler um arquivo (altere o caminho para um válido no seu sistema)
    caminho_arquivo = "C:\\Users\\mathe\\Downloads\\Novo(a) Documento de Texto.txt"
    linhas = ler_arquivo(caminho_arquivo)
    
    print("Conteúdo do arquivo lido:")
    imprimir_lista_de_strings(linhas)

    # Criando autômato
    regexs = ["a", "b", "c"]
    lc_automato = Automato("My Automato", regexs)  # Estado inicial q0

    regexs2 = ["d", "e", "f"]  # Estado q1
    regexs3 = ["g", "h", "i"]  # Estado q2

    lc_automato.add_estado(regexs2, TipoEstado.NORMAL)
    lc_automato.add_estado(regexs3, TipoEstado.FINAL)

    lc_automato.add_transicao_entre_estados("q0", "q1", ["z"])
    lc_automato.add_transicao_entre_estados("q1", "q2", ["z"])

    print(lc_automato.transitar_entre_estados("z"))
    print(lc_automato.transitar_entre_estados("z"))


if __name__ == "__main__":
    main()
